config.org
#+title: Config
#+author: Troy Fischer
#+startup: fold

* UI & Appearance
** Basic Settings
#+begin_src emacs-lisp
(setq user-full-name "Troy Fischer"
      user-mail-address "troytfischer@gmail.com"
      display-line-numbers-type t
      doom-theme 'doom-one
      password-cache-expiry nil   ;; only type password once
      delete-by-moving-to-trash t ;; move files to trash instead of perma deletion
      )

(setq menu-bar-mode -1) ;; otherwise it comes up in terminal mode
;; display time in modeline
(setq display-time-format "%m-%d-%y %l:%M %p"
      display-time-default-load-average nil)
(display-time-mode)

;; line numbers in all modes
(global-display-line-numbers-mode)

;; underline current cursor line
(defun +troy/set-underline ()
  (global-hl-line-mode 1)
  (setq doom--hl-line-mode t)
  (set-face-attribute hl-line-face nil :underline t)
  (set-face-background 'hl-line nil)
  (set-face-underline 'hl-line t)
  (set-face-attribute 'hl-line nil :inherit nil))
(defun +troy/load-theme ()
  (interactive)
  (call-interactively 'consult-theme))
(map! :leader :desc "load-theme" "h t" #'+troy/load-theme)
; (+troy/set-underline)
(add-hook! (prog-mode text-mode conf-mode special-mode feature-mode) #'hl-line-mode)
#+end_src
** Fonts
*Code*
+ Cascadia Code
+ Hack
+ Fira Code
+ Jetbrains Mono
+ Overpass Mono
+ IBM Plex Mono
+ Source Code Pro
+ Monaspace Argon
+ Monaspace Krypton
+ Monaspace Neon
+ Monaspace Radon
+ Monaspace Xenon
+ Monofoki
+ Ubuntu Mono
+ SF Mono
+ Paper Mono
*Variable*
+ Mona Sans
+ Hubot Sans

#+name: main-font
: Overpass Mono

#+name: variable-font
: Crimson Pro

#+begin_src emacs-lisp :var main-font=main-font variable-font=variable-font :results none
(setq +troy/mono-fonts '("Cascadia Code"
                         "Hack"
                         "Fira Code"
                         "Jetbrains Mono"
                         "IBM Plex Mono"
                         "Monaspace Argon"
                         "Monaspace Krypton"
                         "Monaspace Neon"
                         "Monaspace Radon"
                         "Monaspace Xenon"
                         "Monofoki"
                         "Ubuntu Mono"
                         "Fantasque Sans Mono"
                         "SF Mono"
                         "Overpass Mono"
                         "Source Code Pro"
                         "Paper Mono"))

(setq +troy/variable-fonts '("Merriweather"
                             "Inter"
                             "Mona Sans"
                             "Source Sans 3"
                             "IBM Plex Sans"
                             "Crimson Pro"
                             "Hubot Sans"))

(setq +troy/font-types '("Code" "Variable"))

(defun +troy/read-font (fonts)
  (completing-read "Font: " fonts))
(defun +troy/read-face (faces)
  (completing-read "Font Type: " faces))

(defun +troy/change-font ()
  (interactive)
  (let ((face (+troy/read-face +troy/font-types)))
    (let ((font (+troy/read-font (if (string= face "Code") +troy/mono-fonts +troy/variable-fonts))))
      (if (string= face "Code") (+troy/set-mono-font font) (+troy/set-variable-font font))
      (message "%s font set to %s" face font)))
  (doom/reload-font)
  ;; Restart mixed-pitch-mode if it is active (used by Zen/Writeroom) to rebuild face remaps
  (when (bound-and-true-p mixed-pitch-mode)
    (mixed-pitch-mode -1)
    (mixed-pitch-mode 1)))

(defun +troy/set-mono-font (f)
  (setq doom-font (font-spec :family f :size 15)
        doom-big-font (font-spec :family f :size 24)))

(defun +troy/set-variable-font (f)
  (setq doom-variable-pitch-font (font-spec :family f :size 15)))


(+troy/set-mono-font main-font)
(+troy/set-variable-font variable-font)

(map! :leader :desc "change font" "h r F" #'+troy/change-font)
#+end_src

** Dashboard
#+begin_src emacs-lisp
(setq +doom-dashboard-banner-file (expand-file-name "default-emacs.svg" doom-private-dir))  ;; use custom image as banner
#+end_src
* Keybindings & Navigation
** Vertico
#+begin_src emacs-lisp
(map! :map vertico-map "C-l" #'+vertico/enter-or-preview) ;; allow C-l to select an item
#+end_src
** Yank Commands
#+begin_src emacs-lisp
(map! :leader
      (:prefix-map
       ("y" . "yank")
       :desc "org-yank-link" "o" #'link-hint-copy-link-at-point
       :desc "magit-yank-link" "g" #'forge-copy-url-at-point-as-kill))
#+end_src
** Avy
~avy-goto-char-timer~ is a go to movement command (no pun intended)
#+begin_src emacs-lisp
(use-package! avy
  :config
  (map! :leader "s c" #'avy-goto-char-timer)
  (setq avy-all-windows t))
#+end_src
** Ace Window
#+begin_src emacs-lisp
(when (not (bound-and-true-p evil-state))
  (map! :leader
        :desc "split horizontal" "w v" #'split-window-horizontally
        :desc "split vertical" "w s" #'split-window-vertically
        :desc "ace window" "w w" #'ace-window
        :desc "ace window delete" "w d" #'ace-delete-window))
(when (bound-and-true-p evil-state)
  (use-package! ace-window
    :config
    (map! :leader :desc "ace window" "w C-w" #'ace-window)))
#+end_src
** Swiper
Preferable to both doom emac's ~(+default/search-buffer)~ and ~isearch~.
#+begin_src emacs-lisp
(use-package! ivy
  :config
  (when (modulep! :editor evil)
    (map! :map ivy-minibuffer-map
          "C-j" #'ivy-next-line
          "C-k" #'ivy-previous-line)))

(use-package! swiper
  :config
  (map! :leader "s s" #'swiper-isearch)
  (map! "C-s" #'swiper-isearch
        "C-S-s" #'swiper-isearch-backward))
#+end_src
** Pulsar

#+begin_src emacs-lisp
(use-package! pulsar
  :config
  (setq pulsar-pulse-on-window-change t)
  (add-to-list 'pulsar-pulse-functions 'ace-window)
  (pulsar-global-mode 1))
#+end_src
* Editor Behavior
** Evil
#+begin_src emacs-lisp
;; don't yank the text I just pasted over
(setq evil-kill-on-visual-paste nil)
#+end_src
** Dired
#+begin_src emacs-lisp
(use-package! dired-x
  :config
  (setq dired-omit-files (concat dired-omit-files "\\|^\\..+$") ;; hides dotfiles
        dired-omit-files (concat dired-omit-files "\\|__pycache__") ;; hides __pycache__
        dired-deletion-confirmer #'y-or-n-p
        dired-open-extensions '(("mkv" . "mpv")
                                ("mp4" . "mpv"))))
#+end_src
** Feature Mode (Cucumber)
#+begin_src emacs-lisp
(use-package! feature-mode
  :config
  (add-to-list 'auto-mode-alist '("\.feature$" . feature-mode)))
#+end_src
** JSON Mode
#+begin_src emacs-lisp
(add-hook 'json-mode-hook (lambda ()
                            (make-local-variable 'js-indent-level)
                            (setq js-indent-level 2)))
#+end_src
** YAML Mode
#+begin_src emacs-lisp
(add-to-list 'auto-mode-alist '("Bogiefile" . yaml-mode))
#+end_src
** PDF View
#+begin_src emacs-lisp
(defun +troy/pdf-view-config ()
  (display-line-numbers-mode -1)
  (global-hl-line-mode -1))
(add-hook! 'pdf-view-mode-hook #'+troy/pdf-view-config)
#+end_src
* Programming Languages
** LSP Configuration
*** Breadcrumbs
#+begin_src emacs-lisp
(use-package! lsp-mode
  :config
  (setq lsp-headerline-breadcrumb-enable t)
  (map! :leader :after lsp-mode "c R" #'lsp-workspace-restart))
#+end_src
** Python
*** run-python shell
#+begin_src emacs-lisp
(setq python-shell-completion-native-enable nil)
#+end_src
*** LSP
#+begin_src emacs-lisp
(use-package! lsp-diagnostics
  :after flycheck
  :config
  (lsp-diagnostics-flycheck-enable))
#+end_src
**** Pyright
:PROPERTIES:
:ID:       b1a15861-4228-4df5-9552-8f4975a59314
:END:
#+begin_src emacs-lisp
(use-package! lsp-pyright
  :hook
  ((python-mode . (lambda ()
                    (lsp-deferred)))
   (flycheck-mode . (lambda ()
                      (flycheck-add-next-checker 'lsp
                                                 ;; python-pyright is redundant since I use pyright as my lsp server & lsp diagnostics
                                                 ;; 'python-pyright
                                                 'python-mypy))))
  ;;(flycheck-add-next-checker 'python-flake8 'python-pylint)
  :config
  (setq lsp-pyright-type-checking-mode "all"
        lsp-pyright-multi-root nil
        lsp-pyright-langserver-command "basedpyright" ; trying basedpyright for a while
        ))
#+end_src

***** Basedpyright
Notes on ~basedpyright~ setup which may be generally applicable...

As far as I can tell =lsp-pyright.el= is a little bit of a pain to get working with basedpyright. [[https://github.com/emacs-lsp/lsp-pyright/blob/73377169beff8fe22cc6d52d65099db88bf49679/lsp-pyright.el#L294-L296][:new-connection]] is defined as
#+begin_src emacs-lisp
:new-connection (lsp-stdio-connection (lambda ()
                                          (cons (lsp-package-path 'pyright)
                                                lsp-pyright-langserver-command-args)))
#+end_src

when pyright is installed via lsp, ~(lsp-package-path 'pyright)~ gives =~/.config/emacs/.local/etc/lsp/npm/pyright/bin/pyright-langserver=, and it seems to /always/ use that to start the connection no matter what the value of ~lsp-pyright-langserver-command~ is.

In order to get this working I've changed the new-connection function to
#+begin_src emacs-lisp
:new-connection (lsp-stdio-connection (lambda ()
                                          (cons "basedpyright"
                                                lsp-pyright-langserver-command-args)))
#+end_src

then ~straight-rebuild-package~ for lsp-pyright and that finally works (with basedpyright available from PATH ofc)

I've tracked this down multiple times now so I'm documenting for posterity.

***** LSP PYRIGHT MULTIROOT ISSUE
When pyright has ~lsp-pyright-multi-root~ set to ~t~ it runs a single server for
every python project it initializes in. On MacOS this has given me trouble with
~too many open files~ issues. To prevent that I've set the variable to nil in [[file:~/.config/emacs/modules/lang/python/config.el::setq lsp-pyright-multi-root
nil][.config/emacs/modules/lang/python/config.el]].

After setting the variable delete the lsp-session file in
[[file:~/.config/emacs/.local/cache/]] (or delete the entire directory which will have
other repercussions like needing to re-add projects to projectile)
***** Necessary programs
****** Pyright
Doom emacs will prompt to install this server automatically but if that doesn't work...
#+begin_src sh :tangle no
pip3 install -U pyright
#+end_src
****** Flycheck Checkers
Need to install ~pylint~ and ~flake8~ for flycheck.
#+begin_src sh :tangle no
brew install flake8 pylint
#+end_src
****** Ruff code formatter
#+begin_src emacs-lisp
(defun +troy/python-formatter-hook ()
  (setq apheleia-formatter 'ruff)
  ;(setq flycheck-python-ruff-executable "/opt/homebrew/bin/ruff")
  )
(add-hook! python-mode #'+troy/python-formatter-hook)
#+end_src
*** Debugging
=debugpy= is recommended
#+begin_src sh :tangle no
pip3 install -U debugpy ptvsd pytest behave
#+end_src

#+begin_src emacs-lisp
(after! dap-mode
  (setq dap-python-executable "python3")
  (setq dap-python-debugger 'debugpy)
  (require 'dap-python))
(defun refresh-breakpoints ()
  (interactive)
  (set-window-buffer nil (current-buffer)))
#+end_src
**** Key Mapping
#+begin_src emacs-lisp
(map! :leader
      (:prefix-map ("d" . "debug")
       :desc "dap-breakpoint-toggle" "t" #'dap-breakpoint-toggle
       :desc "dap-debug" "d" #'dap-debug
       :desc "dap-debug-recent" "r" #'dap-debug-recent
       :desc "dap-debug-last" "l" #'dap-debug-last
       :desc "dap-debug-edit-template" "e" #'dap-debug-edit-template
       :desc "dap-next" "n" #'dap-next
       :desc "refresh breakpoints" "R" #'refresh-breakpoints
       :desc "disconnect" "q" #'dap-disconnect
       (:prefix-map ("u" . "ui")
        :desc "dap-ui-breakpoints-list" "l" #'dap-ui-breakpoints-list
        :desc "dap-ui-breakpoints-delete" "d" #'dap-ui-breakpoints-delete)))
#+end_src
**** Issue with dap-python
[[file:~/.config/emacs/.local/straight/repos/dap-mode/dap-python.el::let* ((python-executable (dap-python--pyenv-executable-find dap-python-executable)][Link to file on system]]
I have been changing this line: https://github.com/emacs-lsp/dap-mode/blob/c72ac1b1a5811cfda3c6740bcf24c67222b6c840/dap-python.el#L174 to
#+begin_src emacs-lisp :tangle no
(let* ((python-executable (executable-find dap-python-executable))
#+end_src
Because in general I'm using ~pipenv~ rather than ~pyvenv~
**** Custom Dap Template
***** Behave
#+begin_src emacs-lisp :tangle no
(defun +troy/add-behave-template ()
  (interactive)
  (dap-register-debug-template
   "Python :: Run behave (buffer)"
   (list :type "python"
         :args nil
         :cwd (doom-project-root)
         :program nil
         :module "behave"
         :request "launch"
         :name "Python :: Run behave (buffer)")))
#+end_src
*** Formatting
**** Necessary programs
#+begin_src sh :tangle no
brew install ruff
#+end_src
**** Optimize Imports
#+begin_src emacs-lisp :tangle no
(use-package! py-isort
  :after python
  :hook ((before-save . py-isort-before-save)))
#+end_src
***** Pip
#+begin_src sh :tangle no
pip3 install -U pyflakes isort
#+end_src
***** Homebrew
#+begin_src sh :tangle no
brew install isort flake8
#+end_src
*** Add Pip executables to path
#+begin_src emacs-lisp
(let ((pip-path (concat (shell-command-to-string "echo -n $(python3 -m site --user-base)") "/bin")))
  (add-to-list 'exec-path pip-path))
#+end_src
** Lua
Requires my neomvim config and using mason to install the server...


With regard to ~lsp-lua-workspace-library~ I had the same question as [[https://github.com/emacs-lsp/lsp-mode/issues/2818][this guy]] and am equally dumbfounded.
#+begin_src emacs-lisp
(use-package! lsp-lua
  :config
  (setq lsp-clients-lua-language-server-bin (expand-file-name "~/.local/share/nvim/mason/bin/lua-language-server")
        lsp-clients-lua-language-server-main-location (expand-file-name "~/.local/share/nvim/mason/packages/lua-language-server/libexec/main.lua")
        lsp-lua-workspace-library (ht
                                   ((expand-file-name "~/.local/share/nvim/site/pack/deps/opt") t)
                                   ;; Dynamically find neovim runtime path instead of hardcoding version
                                   ((file-name-directory (directory-file-name (file-name-directory (executable-find "nvim")))) t))))
#+end_src
** C++
#+begin_src emacs-lisp
(use-package! clang-format)
(add-hook! 'c++-mode-hook #'lsp-deferred)
#+end_src
* Org Mode
** Config
#+begin_src emacs-lisp
(use-package! org
  :ensure nil
  :config
  (defvar +troy/main-org-agenda-file (expand-file-name (concat org-directory "/agenda.org")))
  (setq org-directory "~/org/"
        org-agenda-files (list +troy/main-org-agenda-file)
        org-default-notes-file (concat org-directory "notes.org")
        org-agenda-span 30
        org-hide-emphasis-markers t)
  (map! :map org-mode-map
        :localleader "TAB" #'org-toggle-inline-images)
  (add-to-list 'org-refile-targets '(org-default-notes-file :maxlevel . 3))
  ;; disable org mode auto complete suggestions
  (add-hook! 'org-mode-hook #'(lambda () (company-mode -1))))
#+end_src
** Agenda
*** Keybinding to open agenda.org
#+begin_src emacs-lisp
(defun +troy/open-org-agenda ()
  (interactive)
  (find-file +troy/main-org-agenda-file))
(map! :leader
      :desc "Open agenda.org" "o a o" #'+troy/open-org-agenda)
#+end_src
** ob-http
[[https://github.com/zweifisch/ob-http][Repository]]
Package to do Postman like things in org mode.
#+begin_src emacs-lisp
(use-package! ob-http
  :after org-babel
  :ensure nil
  :config
  (add-to-list 'org-babel-load-languages '(http . t)))
#+end_src
** Latex
See [[https://emacs.stackexchange.com/questions/79285/why-is-org-highlight-latex-and-related-messing-with-org-hyperlink-description][link]] for =org-highlight-latex-and-related= details. Without that setting the org syntax highlighting gets messed up.
#+begin_src emacs-lisp
(after! org
  (use-package! ox-extra
    :config
    (ox-extras-activate '(latex-header-blocks ignore-headlines)))

  ;; Import ox-latex to get org-latex-classes and other funcitonality
  ;; for exporting to LaTeX from org
  (use-package! ox-latex
    :init
    ;; code here will run immediately
    :config
    ;; code here will run after the package is loaded
    (setq org-latex-with-hyperref nil) ;; stop org adding hypersetup{author..} to latex export
    ;; (setq org-latex-prefer-user-labels t)
    ;; (setq org-latex-compiler "pdflatex")
    ;; (setq latex-run-command "pdflatex")

    (setq org-highlight-latex-and-related '(script entities))
    ;; deleted unwanted file extensions after latexMK
    (setq org-latex-logfiles-extensions
          (quote ("xdv" "lof" "lot" "tex~" "aux" "idx" "log" "out" "toc" "nav" "snm" "vrb" "dvi" "fdb_latexmk" "blg" "brf" "fls" "entoc" "ps" "spl" "bbl" "xmpi" "run.xml" "bcf" "acn" "acr" "alg" "glg" "gls" "ist")))

    (unless (boundp 'org-latex-classes)
      (setq org-latex-classes nil)))
  )
#+end_src
* Tools & Integrations
** Bitwarden
#+begin_src emacs-lisp
(use-package! bitwarden
  :config
  (setq bitwarden-bw-executable (executable-find "bw")
        bitwarden-user "troytfischer@gmail.com"
        bitwarden-automatic-unlock (lambda ()
                                     (let* ((auth-sources '("~/.authinfo.gpg"))
                                            (matches (auth-source-search :user "bitwarden"
                                                                         :host "pass.troyfischer.net"
                                                                         :require '(:secret)
                                                                         :max 1))
                                            (entry (nth 0 matches))
                                            (secret (plist-get entry :secret)))
                                       (if (functionp secret)
                                           (funcall secret)
                                         secret))))
  ;; Enable bitwarden as an auth-source backend (similar to pass +auth)
  (bitwarden-auth-source-enable)

  (map! :leader
        (:prefix-map ("P" . "Passwords")
         :desc "bitwarden-unlock" "u" #'bitwarden-unlock
         :desc "bitwarden-lock" "L" #'bitwarden-lock
         :desc "bitwarden-list-all" "l" #'bitwarden-list-all
         :desc "bitwarden-getpass" "c" #'bitwarden-getpass
         :desc "bitwarden-get-otp" "o" #'bitwarden-get-otp
         :desc "bitwarden-sync" "s" #'bitwarden-sync
         :desc "bitwarden-refresh-account-names" "r" #'bitwarden-refresh-account-names)))
#+end_src
** Auth Sources
#+begin_src emacs-lisp
(after! auth-source
  (add-to-list 'auth-sources "~/.authinfo.gpg"))
#+end_src
** Projectile
#+begin_src emacs-lisp
(use-package! projectile
  :ensure nil
  :config
  (map! :leader :desc "ripgrep" "p G" #'projectile-ripgrep)
  (map! :leader :desc "PyCharm" "p P" #'+troy/open-project-in-pycharm)
  (map! :leader :desc "Neovim" "p N" #'+troy/open-project-in-neovim)
  (map! :leader :desc "Neovim" "p n" #'+troy/open-file-in-neovim)
  (add-to-list 'projectile-globally-ignored-directories "^\\.venv$"))
#+end_src
** Magit
#+begin_src emacs-lisp
(use-package! magit
  :ensure nil
  :config
  (setq magit-log-margin '(t "%Y-%m-%d %I:%M %p " magit-log-margin-width t 18)))
#+end_src
** Code Review
Config: [[https://github.com/wandersoncferreira/code-review]]
#+begin_src emacs-lisp
(setq code-review-auth-login-marker 'forge)

config.el
(setq user-full-name "Troy Fischer"
      user-mail-address "troytfischer@gmail.com"
      display-line-numbers-type t
      doom-theme 'doom-city-lights
      password-cache-expiry nil   ;; only type password once
      delete-by-moving-to-trash t ;; move files to trash instead of perma deletion
      )

(setq menu-bar-mode -1) ;; otherwise it comes up in terminal mode
;; display time in modeline
(setq display-time-format "%m-%d-%y %l:%M %p"
      display-time-default-load-average nil)
(display-time-mode)

;; line numbers in all modes
(global-display-line-numbers-mode)

;; underline current cursor line
(defun +troy/set-underline ()
  (global-hl-line-mode 1)
  (setq doom--hl-line-mode t)
  (set-face-attribute hl-line-face nil :underline t)
  (set-face-background 'hl-line nil)
  (set-face-underline 'hl-line t)
  (set-face-attribute 'hl-line nil :inherit nil))
(defun +troy/load-theme ()
  (interactive)
  (call-interactively 'consult-theme)
  (+troy/set-underline))
(map! :leader :desc "load-theme" "h t" #'+troy/load-theme)
(+troy/set-underline)
(add-hook! (prog-mode text-mode conf-mode special-mode feature-mode) #'hl-line-mode)

(let ((main-font '"Monaspace Argon")
      (variable-font '"Source Code Pro"))
(setq +troy/mono-fonts '("Cascadia Code"
                         "Hack"
                         "Fira Code"
                         "Jetbrains Mono"
                         "IBM Plex Mono"
                         "Monaspace Argon"
                         "Monaspace Krypton"
                         "Monaspace Neon"
                         "Monaspace Radon"
                         "Monaspace Xenon"))
(setq +troy/variable-fonts '("Overpass" "Source Code Pro"))
(setq +troy/font-types '("Code" "Variable"))

(defun +troy/read-font (fonts)
  (completing-read "Font: " fonts))
(defun +troy/read-face (faces)
  (completing-read "Font Type: " faces))

(defun +troy/change-font ()
  (interactive)
  (let ((face (+troy/read-face +troy/font-types)))
    (let ((font (+troy/read-font (if (string= face "Code") +troy/mono-fonts +troy/variable-fonts))))
      (if (string= face "Code") (+troy/set-mono-font font) (+troy/set-variable-font font))
      (message "%s font set to %s" face font)))
  (doom/reload-font))

(defun +troy/set-mono-font (f)
  (setq doom-font (font-spec :family f :size 15)
        doom-big-font (font-spec :family f :size 24)))

(defun +troy/set-variable-font (f)
  (setq doom-variable-pitch-font (font-spec :family f :size 15)))


(+troy/set-mono-font main-font)
(+troy/set-variable-font variable-font)

(map! :leader :desc "change font" "h r F" #'+troy/change-font)
)

(setq +doom-dashboard-banner-file (expand-file-name "default-emacs.svg" doom-private-dir))  ;; use custom image as banner

(defun +troy/open-project-in-pycharm ()
  (interactive)
  (async-shell-command (format "open -na 'PyCharm CE.app' --args %s" (projectile-project-root)))
  (doom/window-maximize-buffer))

(defun +troy/open-project-in-neovim ()
  (interactive)
  (call-process-shell-command (format "alacritty -e nvim %s" (projectile-project-root)) nil 0))
(defun +troy/open-file-in-neovim ()
  (interactive)
  (call-process-shell-command (format "alacritty -e nvim %s" (buffer-file-name)) nil 0))

(use-package! dired-x
  :config
  (setq dired-omit-files (concat dired-omit-files "\\|^\\..+$") ;; hides dotfiles
        dired-omit-files (concat dired-omit-files "\\|__pycache__") ;; hides __pycache__
        dired-deletion-confirmer #'y-or-n-p
        dired-open-extensions '(("mkv" . "mpv")
                                ("mp4" . "mpv"))))

(use-package! feature-mode
  :config
  (add-to-list 'auto-mode-alist '("\.feature$" . feature-mode)))

(setq python-shell-completion-native-enable nil)

(use-package! lsp-diagnostics
  :after flycheck
  :config
  (lsp-diagnostics-flycheck-enable))

(use-package! lsp-pyright
  :hook
  ((python-mode . (lambda ()
                    (lsp-deferred)))
   (flycheck-mode . (lambda ()
                      (flycheck-add-next-checker 'lsp 'python-pyright 'python-ruff))))
  ;;(flycheck-add-next-checker 'python-flake8 'python-pylint)
  :config
  (setq lsp-pyright-venv-directory "~/.local/share/virtualenvs"
        lsp-pyright-type-checking-mode "all"
        lsp-pyright-multi-root nil
        lsp-pyright-langserver-command "basedpyright-langserver"))

(defun +troy/python-formatter-hook ()
  (setq apheleia-formatter 'ruff)
  (setq flycheck-python-ruff-executable "/opt/homebrew/bin/ruff"))
(add-hook! python-mode #'+troy/python-formatter-hook)

(after! dap-mode
  (setq dap-python-executable "python3")
  (setq dap-python-debugger 'debugpy)
  (require 'dap-python))
(defun refresh-breakpoints ()
  (interactive)
  (set-window-buffer nil (current-buffer)))

(map! :leader
      (:prefix-map ("d" . "debug")
       :desc "dap-breakpoint-toggle" "t" #'dap-breakpoint-toggle
       :desc "dap-debug" "d" #'dap-debug
       :desc "dap-debug-recent" "r" #'dap-debug-recent
       :desc "dap-debug-last" "l" #'dap-debug-last
       :desc "dap-debug-edit-template" "e" #'dap-debug-edit-template
       :desc "dap-next" "n" #'dap-next
       :desc "refresh breakpoints" "R" #'refresh-breakpoints
       :desc "disconnect" "q" #'dap-disconnect
       (:prefix-map ("u" . "ui")
        :desc "dap-ui-breakpoints-list" "l" #'dap-ui-breakpoints-list
        :desc "dap-ui-breakpoints-delete" "d" #'dap-ui-breakpoints-delete)))

(let ((pip-path (concat (shell-command-to-string "echo -n $(python3 -m site --user-base)") "/bin")))
  (add-to-list 'exec-path pip-path))

(use-package! lsp-lua
  :config
  (setq lsp-clients-lua-language-server-bin (expand-file-name "~/.local/share/nvim/mason/bin/lua-language-server")
        lsp-clients-lua-language-server-main-location (expand-file-name "~/.local/share/nvim/mason/packages/lua-language-server/libexec/main.lua")
        lsp-lua-workspace-library (ht
                                   ((expand-file-name "~/.local/share/nvim/site/pack/deps/opt") t)
                                   ((concat (string-trim (shell-command-to-string "brew --prefix") ) "/Cellar/neovim/0.10.2_1/share/nvim/runtime" ) t))))

(use-package! clang-format)
(add-hook! 'c++-mode-hook #'lsp-deferred)

(use-package! lsp-mode
  :config
  (setq lsp-headerline-breadcrumb-enable t)
  (map! :leader :after lsp-mode "c R" #'lsp-workspace-restart))

(use-package! org
  :ensure nil
  :config
  (defvar +troy/main-org-agenda-file (expand-file-name (concat org-directory "/agenda.org")))
  (setq org-directory "~/org/"
        org-agenda-files (list +troy/main-org-agenda-file)
        org-default-notes-file (concat org-directory "notes.org")
        org-agenda-span 30
        org-hide-emphasis-markers t)
  (map! :map org-mode-map
        :localleader "TAB" #'org-toggle-inline-images)
  (add-to-list 'org-refile-targets '(org-default-notes-file :maxlevel . 3))
  ;; disable org mode auto complete suggestions
  (add-hook! 'org-mode-hook #'(lambda () (company-mode -1))))

(defun +troy/open-org-agenda ()
  (interactive)
  (find-file +troy/main-org-agenda-file))
(map! :leader
      :desc "Open agenda.org" "o a o" #'+troy/open-org-agenda)

(use-package! ob-http
  :after org-babel
  :ensure nil
  :config
  (add-to-list 'org-babel-load-languages '(http . t)))

(after! org
  (use-package! ox-extra
    :config
    (ox-extras-activate '(latex-header-blocks ignore-headlines)))

  ;; Import ox-latex to get org-latex-classes and other funcitonality
  ;; for exporting to LaTeX from org
  (use-package! ox-latex
    :init
    ;; code here will run immediately
    :config
    ;; code here will run after the package is loaded
    (setq org-latex-with-hyperref nil) ;; stop org adding hypersetup{author..} to latex export
    ;; (setq org-latex-prefer-user-labels t)

    (setq org-highlight-latex-and-related '(script entities))
    ;; deleted unwanted file extensions after latexMK
    (setq org-latex-logfiles-extensions
          (quote ("xdv" "lof" "lot" "tex~" "aux" "idx" "log" "out" "toc" "nav" "snm" "vrb" "dvi" "fdb_latexmk" "blg" "brf" "fls" "entoc" "ps" "spl" "bbl" "xmpi" "run.xml" "bcf" "acn" "acr" "alg" "glg" "gls" "ist")))

    (unless (boundp 'org-latex-classes)
      (setq org-latex-classes nil)))
  )

(defun org-babel-edit-prep:python (babel-info)
  (setq-local buffer-file-name (->> babel-info caddr (alist-get :tangle)))
  (lsp))

(defun +troy/password-store-dir ()
  (find-file "~/.password-store"))
(defun +troy/git-password-store ()
  (interactive)
  (+troy/password-store-dir)
  (magit))
(use-package! password-store
  :config
  (map! :leader
        (:prefix-map ("P" . "Passwords")
         :desc "password-store-copy" "c" #'password-store-copy
         :desc "password-store-otp-token-copy" "o" #'password-store-otp-token-copy
         :desc "password-store-git" "g" #'+troy/git-password-store
         :desc "password-store" "p" #'pass)))

(use-package! projectile
  :ensure nil
  :config
  (map! :leader :desc "ripgrep" "p G" #'projectile-ripgrep)
  (map! :leader :desc "PyCharm" "p P" #'+troy/open-project-in-pycharm)
  (map! :leader :desc "Neovim" "p N" #'+troy/open-project-in-neovim)
  (map! :leader :desc "Neovim" "p n" #'+troy/open-file-in-neovim)
  (add-to-list 'projectile-globally-ignored-directories "^\\.venv$"))

(map! :map vertico-map "C-l" #'+vertico/enter-or-preview) ;; allow C-l to select an item

(setq org-html-postamble-format
      '(("en" "<p class=\"author\">Author: %a</p><p class=\"date\">Updated: %C</p>")))
(setq org-html-postamble t)
(setq org-html-head-include-default-style nil)
(setq org-publish-project-alist
      '(
        ("blog-html"
         :recursive t
         :base-extension "org"
         :base-directory "~/blog/content"
         :publishing-directory "~/blog/public"
         :publishing-function org-html-publish-to-html
         :section-numbers nil
         :org-html-link-home "/index.html"
         :org-html-link-use-abs-url t
         )
        ("blog-static"
         :recursive t
         :base-directory "~/blog/content/"
         :base-extension "css\\|js\\|png\\|jpg\\|jpeg\\|gif\\|pdf\\|mp3\\|ogg\\|swf\\|ico"
         :publishing-directory "~/blog/public/"
         :publishing-function org-publish-attachment
         )
        ("blog" :components ("blog-html" "blog-static"))))
(defun +troy/publish-blog-remote ()
  (interactive)
  (async-shell-command "rsync -L -e ssh -uvrz ~/blog/public/ root@165.227.115.74:/var/www/html/ --delete --chmod=Du=rwx,Dgo=rx,Fu=rw,Fog=r"))

(setq auth-sources '("~/.authinfo.gpg"))

(setq code-review-auth-login-marker 'forge)
(add-hook 'code-review-mode-hook #'emojify-mode)
(add-hook 'code-review-mode-hook
          (lambda ()
            ;; include *Code-Review* buffer into current workspace
            (persp-add-buffer (current-buffer))))

(setq code-review-lgtm-message "LGTM ✔")

(add-hook 'json-mode-hook (lambda ()
                            (make-local-variable 'js-indent-level)
                            (setq js-indent-level 2)))

(add-to-list 'auto-mode-alist '("Bogiefile" . yaml-mode))

(when (version<= "29" emacs-version)
  (setq pixel-scroll-precision-mode t))

(defun file-notify-rm-all-watches ()
  "Remove all existing file notification watches from Emacs."
  (interactive)
  (maphash
   (lambda (key _value)
     (file-notify-rm-watch key))
   file-notify-descriptors))

(when (not (eq system-type 'darwin))
  (use-package! screenshot))

(defun +troy/pdf-view-config ()
  (display-line-numbers-mode -1)
  (global-hl-line-mode -1))
(add-hook! 'pdf-view-mode-hook #'+troy/pdf-view-config)

(use-package! leetcode
  :config
  (setq leetcode-prefer-language "python3")
  (setq leetcode-prefer-sql "mysql")
  (setq leetcode-save-solutions t)
  (setq leetcode-directory "~/leetcode")
  (add-to-list 'lsp-file-watch-ignored-directories "[/\\\\]leetcode\\'")
  (add-hook 'leetcode-solution-mode-hook
          (lambda() (flycheck-mode -1))))

(map! :leader
      (:prefix-map
       ("y" . "yank")
       :desc "org-yank-link" "o" #'link-hint-copy-link-at-point
       :desc "magit-yank-link" "g" #'forge-copy-url-at-point-as-kill))

(use-package! avy
  :config
  (setq avy-all-windows t))

(map! "C-s" #'swiper-isearch
      "C-S-s" #'swiper-isearch-backward)
(map! :map ivy-minibuffer-map
      "C-j" #'ivy-next-line
      "C-k" #'ivy-previous-line)

(when (not (bound-and-true-p evil-state))
  (map! :leader
        :desc "split horizontal" "w v" #'split-window-horizontally
        :desc "split vertical" "w s" #'split-window-vertically
        :desc "ace window" "w w" #'ace-window
        :desc "ace window delete" "w d" #'ace-delete-window))
(when (bound-and-true-p evil-state)
  (use-package! ace-window
    :config
    (map! :leader :desc "ace window" "w C-w" #'ace-window)))

(map! "C-c w h" #' +hydra/window-nav/body)

(when (not (bound-and-true-p evil-state))
  (map! "C-c s r" #'counsel-mark-ring))

(use-package! magit
  :ensure nil
  :config
  (setq magit-log-margin '(t "%Y-%m-%d %I:%M %p " magit-log-margin-width t 18)))

(put 'aio-defun 'edebug-form-spec
     '(name
       (&optional arg-list)
       body))
